plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMetadata.java

```diff
@@ -327,12 +332,44 @@ public ConnectorMergeTableHandle beginMerge(ConnectorSession session, ConnectorT
         List<JdbcColumnHandle> columns = phoenixClient.getColumns(session, plainTable).stream()
                 .filter(column -> !ROWKEY.equalsIgnoreCase(column.getColumnName()))
                 .collect(toImmutableList());
-        PhoenixOutputTableHandle phoenixOutputTableHandle = (PhoenixOutputTableHandle) beginInsert(session, plainTable, ImmutableList.copyOf(columns), retryMode);
+
+        PhoenixOutputTableHandle insertHandle = (PhoenixOutputTableHandle) beginInsert(session, plainTable, ImmutableList.copyOf(columns), retryMode);
+
+        List<JdbcColumnHandle> primaryKeys = phoenixClient.getPrimaryKeyColumnHandles(session, plainTable);
+
+        ImmutableList.Builder<String> updateColumnNamesBuilder = ImmutableList.builder();
+        ImmutableList.Builder<Type> updateColumnTypesBuilder = ImmutableList.builder();
+        // Phoenix use upsert, so all the columns including primary keys will act as the update
+        updateColumnNamesBuilder.addAll(insertHandle.getColumnNames());
+        updateColumnTypesBuilder.addAll(insertHandle.getColumnTypes());
+        if (insertHandle.rowkeyColumn().isPresent()) {
+            updateColumnNamesBuilder.add(ROWKEY);
+            updateColumnTypesBuilder.add(ROWKEY_COLUMN_HANDLE.getColumnType());
+        }
+        PhoenixOutputTableHandle updateHandle = new PhoenixOutputTableHandle(
+                insertHandle.getRemoteTableName(),
+                updateColumnNamesBuilder.build(),
+                updateColumnTypesBuilder.build(),
+                Optional.empty(),
+                Optional.empty());
+
+        // Delete only requires the primary keys
+        PhoenixOutputTableHandle deleteHandle = (PhoenixOutputTableHandle) beginInsert(session, plainTable, ImmutableList.copyOf(primaryKeys), retryMode);
+
+        // The TupleDomain for building the conjuncts of the primary keys
+        ImmutableMap.Builder<ColumnHandle, Domain> primaryKeysDomainBuilder = ImmutableMap.builder();
+        // This value is to build the TupleDomain, but it won't affect the query field in result of the `DefaultQueryBuilder#prepareDeleteQuery`
+        Domain dummy = Domain.singleValue(BIGINT, 0L);
+        for (JdbcColumnHandle columnHandle : primaryKeys) {
+            primaryKeysDomainBuilder.put(columnHandle, dummy);
+        }
```

Comment:
Why this code changed so much here?

Reply:
Prepared the `insert`, `update`, `delete` handle for the `PhoenixMergeSink` that can use directly. I will remove this change and to see if we actually need this later.